File: csharp/dotnet/pxprpc/BoundMethodCallable .cs
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Text;

namespace pxprpc
{
    public class BoundMethodCallable : MethodCallable
    {
        protected Object boundObj;
        public BoundMethodCallable(MethodInfo method, Object boundObj):base(method)
        {
            this.boundObj = boundObj;
        }
        public override void readParameter(PxpRequest req)
        {
            ServerContext c = req.context;
            Object[] args = new Object[argList.Length];
            if (firstInputParamIndex >= 1)
            {
                args[0] = null;
            }
            for (int i = firstInputParamIndex; i < argList.Length; i++)
            {
                args[i] = readNext(c, argList[i]);
            }
            req.parameter = new Object[] { null, args };
        }
        public override void call(PxpRequest req, Action<Object> asyncRet)
        {
            ServerContext ctx = req.context;

            Object result = null;
            Object[] args = (Object[])((Object[])req.parameter)[1];
            if (firstInputParamIndex >= 1)
            {
                args[0] = asyncRet;
            }
            try
            {
                result = method.Invoke(this.boundObj, args);
                if (firstInputParamIndex == 0)
                {
                    asyncRet(result);
                }
            }
            catch (Exception ex)
            {
                asyncRet(ex.InnerException);
            }

            
        }
    }
}

File: java/src/pxprpc/extend/BoundMethodCallable.java
package pxprpc.extend;

import pxprpc.base.PxpRequest;
import pxprpc.base.ServerContext;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class BoundMethodCallable extends MethodCallable{

	public Object boundObj;
	public BoundMethodCallable(Method method,Object boundObj) {
		super(method);
		this.boundObj=boundObj;
	}

	@Override
	public void parseMethod() {
		Class<?>[] paramsType = method.getParameterTypes();
		if (paramsType.length>0 && paramsType[0] == AsyncReturn.class) {
			firstInputParamIndex = 1;
		}
		if(!parseCustomTypeDecl()){
			char[] t1 = new char[paramsType.length-firstInputParamIndex];
			for(int i=firstInputParamIndex;i<paramsType.length;i++){
				t1[i-firstInputParamIndex]= TypeDeclParser.jtypeToValueInfo(paramsType[i]);
			}
			tParam=t1;
			if(method.getReturnType()!=void.class){
				tResult = new char[]{
						TypeDeclParser.jtypeToValueInfo(method.getReturnType())
				};
			}else{
				tResult=new char[0];
			}
		}
	}

	public void call(final PxpRequest req) {
		ServerContext ctx = req.context;
		try {
			Object[] args=readParameter(req);
			Object[] args2=new Object[firstInputParamIndex+args.length];
			System.arraycopy(args,0,args2,firstInputParamIndex,args.length);
			if(firstInputParamIndex==1){
				args2[0]=new AsyncReturn<Object>() {
					@Override
					public void resolve(Object result) {
						BoundMethodCallable.this.writeResult(req,result);
						req.done();
					}
					@Override
					public void reject(Exception ex) {
						req.rejected=ex;
						req.done();
					}

					@Override
					public PxpRequest getRequest() {
						return req;
					}
				};
			}
			Object result=method.invoke(boundObj,args2);
			if(firstInputParamIndex==0) {
				this.writeResult(req,result);
				req.done();
			}
		} catch (InvocationTargetException e) {
			req.rejected=e.getCause();
			req.done();
		} catch (Exception e) {
			req.rejected=e;
			req.done();
		}
	}

}
